proposals/0047-last-hard-fork-syscall.md

```diff
@@ -0,0 +1,110 @@
+---
+simd: '0047'
+title: Syscall to get the last restart slot
+authors:
+  - Godmode Galactus (Mango Markets)
+category: Standard
+type: Core
+status: Draft
+created: 2023-04-15
+---
+
+## Summary
+
+Create a new syscall which can be used to get the last restart slot (currently
+the last time hard fork was done on the cluster).
+
+`fn sol_get_last_restart_slot() -> Slot`
+
+## Motivation
+
+In Solana, when the cluster cannot reach consensus, it is currently restarted
+using `ledger-tool` with a hard fork on a slot. All participating nodes need to
+restart their validator specifying the hard fork slot. Once all nodes
+participating in the restart effort on the hard fork exceed 80% of the total
+stakes, the restart is successful, and the cluster continues from the hard fork
+slot. So hard fork is a tool to intentionally fork off the nodes not
+participating in the restart effort. Currently can consider the slot on which
+hardfork was done as a restart slot.
+
+Program developers may find it useful to know that a hard fork has recently
+occurred. This information can help them prevent arbitrage and liquidation
+caused by outdated oracle price account states. However, the cluster's restart
+process takes several hours; in that time, the world can change, causing asset
+prices to fluctuate. As the cluster updates the state of all accounts, malicious
+actors could take advantage of the delay by conducting trades using outdated
+prices, anticipating that they will soon catch up to the actual prices of
+real-world assets. Knowing that cluster restart has been done recently, programs
+can manage these cases more appropriately.
+
+## Alternatives Considered
+
+No alternate considerations; we need to have the value of last restart slot
+while executing the program to correctly treat this case. We cannot have an
+account because then it should be updated just after the restart is successful,
+which will add complexity. The best way is to create a new syscall to get this
+information during execution of transaction, which will help us get last
+restart slot without any interface changes for the instruction.
```

Comment:
> The additional complexity of a syscall is worthwhile if this hard fork info needs to be accessed frequently, as loading in another account makes a txn more expensive.

Right, adding a sysvar account takes up tx account space which is something that DeFi programs tend to struggle with, although perhaps less now with LUTs.

The idea here is to do something quick, minimal, and low risk for existing programs to enable some kind of incremental improvement before the next restart (🔥🔥 ).

Generally we'd expect this syscall to be used before every defi op to sanity check the current network state

Reply:
I think we need a sysvar anyway because syscalls also use sysvars under the hood to serve the request. And yes, syscalls will all become instructions of built-in programs. Meaning, we will try to avoid constant functions which don't perform any calculations in the future and use sysvars instead.

Reply:
I agree with mvines on this because with syscall we need not need to change the existing instruction signature to use this feature. Sysvar we have to update the program to pass the new sysvar inorder to use this feature. Already the limited tx account space is an issue. Syscall will really simplify things and serves the purpose for this feature.

Reply:
ok, thanks. Restoring the sysvar seems fine then, and that resolves the concern that Jon had about front-end accessibility

Reply:
Could we have both syscall and sysvar?

Reply:
++,Is it possible for us to incorporate both syscall and sysvar?

Reply:
it's very annoying to retro-actively add a sysvar to a running program, but if the syscall is anyways going away with the new runtime and won't be accessible anymore (if i understood @Lichtso correctly) let's rip that band-aid off now, rather than later no?

Reply:
> Could we have both syscall and sysvar?

Yes, that is probably the way to go.
Though, the syscall will only exist temporarily in the current program runtime and will be deprecated with it.

Reply:
I will update the SIMD accordingly then. Thanks.

Reply:
Done